### DISCUSSION TITLE
Welcome to OpenSMTPD Discussions!

### DISCUSSION THREAD

**Discussion by poolpOrg** [OpenSMTPD/OpenSMTPD#1103] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by PopeNobody** [comment]
Hi, I am the Nobody formerly known as Rich Paul ( http://electnobody.com ... because nobody knows how to live your life better than you do, and nobody should be telling you what to do ), and I happened on this while trying to unravel a mystery.

My opensmtpd is having ssl issues ... apparently doesn't like openssl ... and I am working to recompile it.  Which worked fine, except that now opensmtpd can't see the pki statments in my smtpd.conf, and dies of shame when I try to listen using ssl ...

Which was a mystery.

Then I found a bigger one.

This is the conversation I had with ChatGPT about it ...

https://chat.openai.com/share/3c877b1b-f7b4-427f-a059-4bf01cfe743e

Thought you might be amused.  If you have any idea how this works with gnu make, I'd love to know.  I've straced the hell out of it, and am clueless.

-------

I'm working on code from openbsd.

There is a construct in a makefile:

.include <bsd.subdir.mk>

There does not seem to be any such file anywhere on my system, which laves the makefile bereft of it's most essential organs ... but yet the beast will not die!


-------

Fear not, for the construct you encountered in the OpenBSD makefile is not the result of sorcery, but rather a built-in mechanism in OpenBSD's make utility. The .include <bsd.subdir.mk> is not a file that you're supposed to find in your system, but rather a directive that includes the bsd.subdir.mk makefile fragment provided by OpenBSD's build system.

**Reply by PopeNobody** [reply]
Ok, I found it, the make process never enters the directory I was in, the source just lives there.

**Reply by omar-polo** [reply]
As you've found out, the makefiles used for the portable build are under `mk/`.  These makefiles are automatically generated by automake and so aren't exactly what I'd define a good reading material, but the automake files (the various `Makefile.am`) should be quite readable and it's there where the magic lives.

Could you please expand what you mean with "apparently doesn't like openssl"?  If you can, showing the output from the configure script and the build (including the various flags you passed if any) would be invaluable to troubleshoot this issue.

Thanks!

**Reply by PopeNobody** [reply]
BSD has some really elegant makefiles, from what I have seen.  I downloaded an iso, 'cause I want to know more.  I'm working on a ports style single tree "linux-from-scratch" project, and I want the inspiration.

As for my ssl problems. i was getting some "internal error" and "wrong version number" messages, and incoming mail via ssl and tls were failing.  I've got a file where I used openssl s_client to connect to both opensmtpd and nginx ... the nginx worked, but the opensmtpd did not.

I ended up upgrading to 7.3, from cvs, and had to hack slightly to get it to work.  I had missing symbols when I tried to compile against libressl's libssl, and multiply defined ones when I used their libtls.  My solution was to remove opensmtpd's libtls, in openbsd_compat, and that solved it.  Google is now sending me messages.

They have decided they don't want mine until I have spf and dkim, so I'm working on that now.  spf is failing mysteriously, so that might keep me busy for a while.
